source/agora/script/Engine.d

```diff
@@ -493,6 +494,21 @@ public class Engine
                     return "VERIFY_SIG signature failed validation";
                 break;
 
+            case OP.VERIFY_HEIGHT_LOCK:
+                if (stack.count() < 1)
+                    return "VERIFY_HEIGHT_LOCK opcode requires a lock height on the stack";
+
+                const height_bytes = stack.pop();
+                if (height_bytes.length != ulong.sizeof)
+                    return "VERIFY_HEIGHT_LOCK height lock must be an 8-byte number";
+
+                const Height height_lock = Height(littleEndianToNative!ulong(
+                    height_bytes[0 .. ulong.sizeof]));
+                if (height_lock > tx.height_lock)
```

Comment:
Is the `tx.height_lock` always the height of a proposed block in the situation of OP.VERIFY_HEIGHT_LOCK being used?

Reply:
It's the height of some future block height, for example +10 blocks in the future.

The engine itself validates the height lock in the transaction, while the consensus rules validate that such a transaction won't be added to the blockchain until that height is reached. So a transaction can be valid, just not yet unlocked. 

It allows creating time contracts between multiple parties. The transaction will be valid, but it will be guaranteed that it will not be included in a block until the height in `height_lock` is reached.

Reply:
The chapter on timelock is very useful: https://cypherpunks-core.github.io/bitcoinbook/ch07.html

Reply:
I got it. You mean that the transaction having `height_lock` higher than the current height is just stored in the transaction pool yet before the height is reached.
But is there a kind of double-spending problem? I guess some bad node can make two transactions, one transaction having `tx.height_lock` of current height and the other transaction having `tx_height_lock` of future block height.

Reply:
That's right. And the book mentions that also. That's why input-level locks are better. However, transaction-level locks are still useful for HTLCs. For example https://bitcoin.stackexchange.com/questions/87968/why-do-we-use-cltv-instead-of-csv-in-the-htlc-output-of-the-commitment-transacti and https://en.bitcoin.it/wiki/Hash_Time_Locked_Contracts.

Reply:
The flash layer will mainly use an input-level lock. However, for multi-hop payments we will have to use HTLCs.

Reply:
Ah, I get to know, to some extent, why you introduce the input-level locks. Ty.